### PR TITLE
Feature: Initialize beam from arrays

### PR DESCRIPTION
**Pull request by RTSandberg**
This PR documents how to add a custom beam to an ImpactX simulation.

More specifically, it shows how to add a beam specified by array-like structures.
This allows various applications of interest, 
such as using a beam from a different simulation, 
initializing a beam from file,
or creating a custom distribution.
This example includes a set of utilities for transforming the beam from fixed-t to fixed-s coordinates, suggesting how a beam from a t-based code can be loaded into ImpactX.

The beam example used here is a ring in x-y, Gaussian in px-py, and has a linear correlation in z-pz.

phase_space

### CODE REVIEW

File: examples/initialize_from_array/analyze_from_array.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
+#!/usr/bin/env python3
+#
+# Copyright 2022-2023 ImpactX contributors
+# Authors: Axel Huebl, Chad Mitchell
```
Comment (n01r) [inline examples/initialize_from_array/analyze_from_array.py]: You might want to update the authors, and put yourself there :)

File: examples/initialize_from_array/run_from_array.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+qm_eev = -1.0 / 0.510998950 / 1e6  # electron charge/mass in e / eV
+ref.z = 0
+
+pc = sim.particle_container()
+
+dx, dy, dz, dpx, dpy, dpz = pycoord.to_ref_part_t_from_lab_t(ref, x, y, z, px, py, pz)
+dx, dy, dt, dpx, dpy, dpt = pycoord.to_s_from_t(ref, dx, dy, dz, dpx, dpy, dpz)
+
+if not on_gpu:
+    dx_podv = amr.PODVector_real_std()
```
Comment (n01r) [inline examples/initialize_from_array/run_from_array.py:65]: Perhaps a comment above what this does?

File: examples/initialize_from_array/transformation_utilities.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
+import numpy as np
+
+
+def to_ref_part_t_from_lab_t(ref_part, x, y, z, px, py, pz):
```
Comment (n01r) [inline examples/initialize_from_array/transformation_utilities.py]: You could add docstrings for the functions here so they show up when used in Jupyter notebooks for instance :)

File: examples/initialize_from_array/visualize_from_array.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+)
+impactx_surrogate_steps = list(beam_impactx_surrogate_series.iterations)
+impactx_surrogate_ref_particle = ix_plt.read_time_series("diags/ref_particle.*")
+
+millimeter = 1.0e3
+micron = 1.0e6
+step = 1
+beam_at_step = beam_impactx_surrogate_series.iterations[step].particles["beam"].to_df()
+ref_part_step = impactx_surrogate_ref_particle.loc[step]
+ref_part = myref = coord.MyRefPart(
```
Comment (n01r) [inline examples/initialize_from_array/visualize_from_array.py]: `myref` is unused

File: examples/initialize_from_array/visualize_from_array.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+ax.set_xlabel("x (mm)")
+ax.set_ylabel("y (mm)")
+# ax = axT[0][1]
+# ax.hist2d(z*millimeter,x*millimeter,bins=200);
+# ax.set_xlabel('z (mm)')
+# ax.set_ylabel('x (mm)')
+# ax = axT[0][2]
+# ax.hist2d(z*millimeter,y*millimeter,bins=200);
+# ax.set_xlabel('z (mm)')
+# ax.set_ylabel('y (mm)')
```
Comment (n01r) [inline examples/initialize_from_array/visualize_from_array.py]: Please remove commented code that is not used or comment in which case people can comment it in :)

**Review by n01r — COMMENTED** [review]
Thanks a lot @RTSandberg for adding this functionality! :sparkles: 
I added my small comments :)

File: examples/initialize_from_array/analyze_from_array.py (hunk 1 above)
Comment (RTSandberg) [reply examples/initialize_from_array/analyze_from_array.py]: Yes, that shows some bad copy-pasting habits :)

Comment (n01r) [reply examples/initialize_from_array/analyze_from_array.py]: No worries, we all do it :)

File: examples/initialize_from_array/visualize_from_array.py (hunk 5 above)
Comment (RTSandberg) [reply examples/initialize_from_array/visualize_from_array.py]: Yep, that is vestigial

File: examples/initialize_from_array/visualize_from_array.py (hunk 4 above)
Comment (RTSandberg) [reply examples/initialize_from_array/visualize_from_array.py]: good catch

File: examples/initialize_from_array/analyze_from_array.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+#
+
+import impactx_read_utilities as ix_read
+import numpy as np
+import openpmd_api as io
+import transformation_utilities as coord
+
+print("Initial Beam:")
+
+beam_impactx_surrogate_series = io.Series(
```
Comment (ax3l) [inline examples/initialize_from_array/analyze_from_array.py]: maybe rename `beam_impactx_surrogate_series` with `beam_series`?

File: examples/initialize_from_array/analyze_from_array.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    "diags/openPMD/monitor.bp", io.Access.read_only
+)
+impactx_surrogate_steps = list(beam_impactx_surrogate_series.iterations)
```
Comment (ax3l) [inline examples/initialize_from_array/analyze_from_array.py]: rename to `impactx_steps`?

**Review by ax3l — COMMENTED** [review]
Please add the test to the end of `examples/CMakeLists.txt` so it is run with `ctest` and in CI.

**Review by cemitch99 — COMMENTED** [review]
Thanks a lot; this is a great feature to have!  I'm not sure how much control we would like to give to users to add/subtract the particle and reference phase space values--the particle coordinates and the reference particle use different coordinate systems, and the momenta have different normalization conventions, so this can easily lead to confusion.  In most cases, working with the local particle data is enough.  (Coupling to WarpX is obviously an exception.)  This might be worth some follow-up discussion before merging.

**Comment by RTSandberg** [comment]
Ok, I put this on the agenda for tomorrow's meeting to discuss

File: examples/initialize_from_array/run_from_array.py [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+for element in dpy:
+    dpy_podv.push_back(element)
+for element in dpt:
+    dpt_podv.push_back(element)
+
+pc.add_n_particles(
+    dx_podv, dy_podv, dt_podv, dpx_podv, dpy_podv, dpt_podv, qm_eev, bunch_charge_C
+)
+
+monitor = elements.BeamMonitor("monitor")
```
Comment (ax3l) [inline examples/initialize_from_array/run_from_array.py]:
```suggestion
monitor = elements.BeamMonitor("monitor", backend="h5")
```

File: examples/initialize_from_array/analyze_from_array.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+#
+
+import impactx_read_utilities as ix_read
+import numpy as np
+import openpmd_api as io
+import transformation_utilities as coord
+
+print("Initial Beam:")
+
+beam_series = io.Series("diags/openPMD/monitor.bp", io.Access.read_only)
```
Comment (ax3l) [inline examples/initialize_from_array/analyze_from_array.py]:
```suggestion
beam_series = io.Series("diags/openPMD/monitor.h5", io.Access.read_only)
```

File: examples/initialize_from_array/visualize_from_array.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+#
+# -*- coding: utf-8 -*-
+
+import impactx_read_utilities as ix_plt
+import openpmd_api as io
+import transformation_utilities as coord
+from matplotlib import pyplot as plt
+
+######## plot phase spaces ###########
+beam_series = io.Series("diags/openPMD/monitor.bp", io.Access.read_only)
```
Comment (ax3l) [inline examples/initialize_from_array/visualize_from_array.py]:
```suggestion
beam_series = io.Series("diags/openPMD/monitor.h5", io.Access.read_only)
```

File: examples/initialize_from_array/run_from_array.py [hunk 11]
```diff
@@ -0,0 +1,108 @@
+import numpy as np
+
+try:
+    import torch
+
+    on_gpu = torch.cuda.is_available()
```
Comment (ax3l) [inline examples/initialize_from_array/run_from_array.py]: Using `torch` for this is a wildly huge dependency :) I think you meant to use
```py
from impactx import Config, ImpactX, elements
```
and then use `Config.have_gpu`.

File: examples/initialize_from_array/run_from_array.py [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+ref.z = 0
+
+pc = sim.particle_container()
+
+dx, dy, dz, dpx, dpy, dpz = pycoord.to_ref_part_t_from_global_t(
+    ref, x, y, z, px, py, pz
+)
+dx, dy, dt, dpx, dpy, dpt = pycoord.to_s_from_t(ref, dx, dy, dz, dpx, dpy, dpz)
+
+if not on_gpu:  # initialize using cpu-based PODVectors
```
Comment (ax3l) [inline examples/initialize_from_array/run_from_array.py]: Do we we need this differentiation or will `PODVector_real_arena` work in both cases?

File: examples/initialize_from_array/run_from_array.py [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+else:  # initialize on device using arena/gpu-based PODVectors
+    dx_podv = amr.PODVector_real_arena()
+    dy_podv = amr.PODVector_real_arena()
+    dt_podv = amr.PODVector_real_arena()
+    dpx_podv = amr.PODVector_real_arena()
+    dpy_podv = amr.PODVector_real_arena()
+    dpt_podv = amr.PODVector_real_arena()
+
+for element in dx:
+    dx_podv.push_back(element)
```
Comment (ax3l) [inline examples/initialize_from_array/run_from_array.py]: Does this truly work if run without managed memory and with a GPU build?

File: examples/initialize_from_array/run_from_array.py [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+    dpt_podv = amr.PODVector_real_std()
+else:  # initialize on device using arena/gpu-based PODVectors
+    dx_podv = amr.PODVector_real_arena()
+    dy_podv = amr.PODVector_real_arena()
+    dt_podv = amr.PODVector_real_arena()
+    dpx_podv = amr.PODVector_real_arena()
+    dpy_podv = amr.PODVector_real_arena()
+    dpt_podv = amr.PODVector_real_arena()
+
+for element in dx:
```
Comment (ax3l) [inline examples/initialize_from_array/run_from_array.py]: The name `element` is not ideal, as it can be confused with beamline elements below. Maybe name it `p_dx` or so.

File: examples/initialize_from_array/run_from_array.py (hunk 12 above)
Comment (RTSandberg) [reply examples/initialize_from_array/run_from_array.py]: No, when I tested this on CPU it didn't work. Maybe there is some other amrex option that needs to be altered

File: examples/initialize_from_array/run_from_array.py (hunk 13 above)
Comment (RTSandberg) [reply examples/initialize_from_array/run_from_array.py]: I haven't tested that combination yet

File: examples/initialize_from_array/README.rst [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+
+In this example, a custom beam is specified at fixed t, transformed to fixed s, and
+then loaded in ImpactX. The custom beam is a ring in x-y,
+with radius r=2 mm,
+radial width :math:`\sigma_r = 5\ \mu`m;
+Gaussian in :math:`p_x` and :math:`p_y` with momentum width :math:`\sigma_p=10`;
+and chirped in z-pz with bunch length :math:`\sigma_z=1` mm,
+mean energy about 10 GeV, 1% uncorrelated energy spread, and z-pz covariance of -0.18.
+
+The transformation utility function reproduces the t-to-s and s-to-t transformations
```
Comment (ax3l) [inline examples/initialize_from_array/README.rst]: @RTSandberg can you please document the assumptions (e.g., beam straight along z) of this example, as we discussed with @cemitch99, in this Readme? :pray:

File: examples/initialize_from_array/run_from_array.py (hunk 12 above)
Comment (ax3l) [reply examples/initialize_from_array/run_from_array.py]: Ok, then leave this for now if it works that way.

File: examples/initialize_from_array/run_from_array.py (hunk 13 above)
Comment (ax3l) [reply examples/initialize_from_array/run_from_array.py]: non-managed memory is the default now, so you just need to test it on GPU.

File: examples/initialize_from_array/run_from_array.py [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+#
+# -*- coding: utf-8 -*-
+
+import amrex.space3d as amr
+from impactx import ImpactX, elements
+from impactx import Config, ImpactX, elements
+import numpy as np
+import transformation_utilities as pycoord
+
+on_gpu = Config.have_gpu
```
Comment (ax3l) [inline examples/initialize_from_array/run_from_array.py]: You can remove the temporary variable and use `Config.have_gpu` below directly.
```suggestion
```

File: examples/CMakeLists.txt [hunk 17]
```diff
@@ -906,3 +906,13 @@ add_impactx_test(triplet.py
     examples/optimize_triplet/plot_triplet.py
 )
 label_impactx_test(triplet.py slow)
+# Initialize From Array ###########################################################
```
Comment (ax3l) [inline examples/CMakeLists.txt]:
```suggestion

# Initialize a Beam from Arrays ##############################################
```

File: examples/initialize_from_array/README.rst [hunk 18]
```diff
@@ -0,0 +1,85 @@
+.. _examples-from-array:
+
+Initialize a beam from arrays
+=============================
```
Comment (ax3l) [inline examples/initialize_from_array/README.rst]:
```suggestion
Initialize a Beam from Arrays
=============================
```

File: examples/initialize_from_array/impactx_read_utilities.py [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+    return pd.concat(
+        (
+            pd.read_csv(filename, delimiter=r"\s+")
+            for filename in glob.glob(file_pattern)
+        ),
+        axis=0,
+        ignore_index=True,
+    ).set_index("id")
+
+
```
Comment (ax3l) [inline examples/initialize_from_array/impactx_read_utilities.py]: unused
```suggestion
```

File: examples/initialize_from_array/transformation_utilities.py [hunk 20]
```diff
@@ -0,0 +1,193 @@
...
+    """
+    Transform from global coordinates to reference particle coordinates
+
+    This function takes particle bunch data and returns the bunch phase space positions relative to a reference particle.
+    It is assumed that the local and global coordinate frames align.
+    That is, the beam transverse directions x and y are the global x and y directions
+    and the beam z direction is the global z direction.
+
+    Parameters
+    ---
```
Comment (ax3l) [inline examples/initialize_from_array/transformation_utilities.py]:
```suggestion
    Parameters
    ----------
```

File: examples/initialize_from_array/transformation_utilities.py [hunk 21]
```diff
@@ -0,0 +1,193 @@
...
+    Transform from reference particle to global coordinates.
+
+    This function takes particle bunch data relative to a reference particle
+    and returns all particle data in the global coordinate frame.
+    It is assumed that the local and global coordinate frames align.
+    That is, the beam transverse directions x and y are the global x and y directions
+    and the beam z direction is the global z direction.
+
+    Parameters
+    ---
```
Comment (ax3l) [inline examples/initialize_from_array/transformation_utilities.py]:
```suggestion
    Parameters
    ----------
```

File: examples/initialize_from_array/transformation_utilities.py [hunk 22]
```diff
@@ -0,0 +1,193 @@
...
+def to_s_from_t(ref_part, dx, dy, dz, dpx, dpy, dpz):  # data_arr_t):
+    """
+    Transform from fixed-s to fixed-t coordinates
+
+    This function takes particle bunch data relative to a reference particle
+    at a fixed time t and returns data at a fixed longitudinal position s.
+    That is, spatial distance is transformed to time delay from the reference particle.
+
+    Parameters
+    ---
```
Comment (ax3l) [inline examples/initialize_from_array/transformation_utilities.py]:
```suggestion
    Parameters
    ----------
```

File: examples/initialize_from_array/transformation_utilities.py [hunk 23]
```diff
@@ -0,0 +1,193 @@
...
+def to_t_from_s(ref_part, dx, dy, dt, dpx, dpy, dpt):  # data_arr_t):
+    """
+    Transform from fixed-t to fixed-s coordinates
+
+    This function takes particle bunch data relative to a reference particle
+    at a fixed longitudinal position s and returns data at a fixed time t.
+    That is, time delay is transformed to spatial distance from the reference particle.
+
+    Parameters
+    ---
```
Comment (ax3l) [inline examples/initialize_from_array/transformation_utilities.py]:
```suggestion
    Parameters
    ----------
```

File: examples/initialize_from_array/README.rst [hunk 24]
```diff
@@ -0,0 +1,85 @@
...
+
+This example demonstrates how a beam can be initalized in ImpactX from array-like structures.
```
Comment (ax3l) [inline examples/initialize_from_array/README.rst:11]:
```suggestion

.. note::

   This example is an early draft of this workflow and will be simplified in future versions of ImpactX, adding direct support for NumPy and CuPy arrays.

This example demonstrates how a beam can be initalized in ImpactX from array-like structures.
```

File: examples/initialize_from_array/run_from_array.py [hunk 25]
```diff
@@ -0,0 +1,110 @@
...
+ref.z = 0
+
+pc = sim.particle_container()
+
+dx, dy, dz, dpx, dpy, dpz = pycoord.to_ref_part_t_from_global_t(
+    ref, x, y, z, px, py, pz
+)
+dx, dy, dt, dpx, dpy, dpt = pycoord.to_s_from_t(ref, dx, dy, dz, dpx, dpy, dpz)
+
+if not on_gpu:  # initialize using cpu-based PODVectors
```
Comment (ax3l) [inline examples/initialize_from_array/run_from_array.py]:
```suggestion
if not Config.have_gpu:  # initialize using cpu-based PODVectors
```

**Review by ax3l — APPROVED** [review]
Thank you, Ryan! :sparkles:
